clang/lib/Driver/Driver.cpp

```diff
@@ -3854,24 +3854,35 @@ class OffloadingActionBuilder final {
     getDeviceDependences(OffloadAction::DeviceDependences &DA,
                          phases::ID CurPhase, phases::ID FinalPhase,
                          PhasesTy &Phases) override {
+      bool SYCLDeviceOnly = Args.hasArg(options::OPT_fsycl_device_only);
       if (CurPhase == phases::Preprocess) {
         // Do not perform the host compilation when doing preprocessing only
         // with -fsycl-device-only.
         bool IsPreprocessOnly =
             Args.getLastArg(options::OPT_E) ||
             Args.getLastArg(options::OPT__SLASH_EP, options::OPT__SLASH_P) ||
             Args.getLastArg(options::OPT_M, options::OPT_MM);
-        if (Args.hasArg(options::OPT_fsycl_device_only) && IsPreprocessOnly) {
-          for (Action *&A : SYCLDeviceActions)
-            A = C.getDriver().ConstructPhaseAction(C, Args, CurPhase, A,
-                                                   AssociatedOffloadKind);
-          return ABRT_Ignore_Host;
+        if (IsPreprocessOnly) {
+          for (Action *&A : SYCLDeviceActions) {
+            if (!SYCLDeviceOnly) {
+              A = C.getDriver().ConstructPhaseAction(C, Args, CurPhase, A,
+                                                     AssociatedOffloadKind);
+              // Add an additional compile action to generate the integration
+              // header.
+              Action *CompileAction =
+                  C.MakeAction<CompileJobAction>(A, types::TY_Nothing);
+              DA.add(*CompileAction, *ToolChains.front(), nullptr,
+                     Action::OFK_SYCL);
+            } else
+              A = C.getDriver().ConstructPhaseAction(C, Args, CurPhase, A,
+                                                     AssociatedOffloadKind);
+          }
```

Comment:
Would something like this preserve the functional meaning?
The early exit bit feels optional.
```suggestion
            A = C.getDriver().ConstructPhaseAction(C, Args, CurPhase, A,
                                                   AssociatedOffloadKind);
            if (SYCLDeviceOnly)
              continue;
            // Add an additional compile action to generate the integration
            // header.
            Action *CompileAction =
                C.MakeAction<CompileJobAction>(A, types::TY_Nothing);
            DA.add(*CompileAction, *ToolChains.front(), nullptr,
                   Action::OFK_SYCL);
          }
```